feat: [FC-0070] rendering split test content in unit page

ihor-romaniuk opened the pull request:
## Description
This PR introduces functionality to display Split Test Content within the new Unit page interface.

The feature enables opening a Split Test Content page within the Unit page’s new interface. This page displays the xBlocks from the specified groups and provides basic configuration options for the groups.

https://github.com/user-attachments/assets/61166a12-55f1-4b4a-837e-d9e4cf45f4ad

## Related Pull Requests
- https://github.com/openedx/edx-platform/pull/35813 - events to trigger click on View button and styles

## Testing instructions
- Create a new course.
- Create several units.
- Open the course unit page.
- Add a Content Experiment xBlock to the unit.
- Click the “View” button on the Content Experiment xBlock.

## Other information
**Settings**
```
EDX_PLATFORM_REPOSITORY: https://github.com/raccoongang/edx-platform.git
EDX_PLATFORM_VERSION: romaniuk/children-container-for-rendering-split-test

TUTOR_GROVE_NEW_MFES:
  authoring:
    port: 18000
    repository: https://github.com/raccoongang/frontend-app-course-authoring.git
    version: romaniuk/children-container-for-rendering-split-test

TUTOR_GROVE_WAFFLE_FLAGS:
  - name: contentstore.new_studio_mfe.use_new_unit_page
    everyone: true

TUTOR_GROVE_MFE_LMS_COMMON_SETTINGS:
  MFE_CONFIG:
    ENABLE_UNIT_PAGE: true
```

openedx-webhooks commented:
Thanks for the pull request, @ihor-romaniuk!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->


#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/2u-tnl`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

PKulkoRaccoonGang commented on src/course-unit/data/thunk.js:
```diff
@@ -202,18 +201,27 @@ export function createNewCourseXBlock(body, callback, blockId, sendMessageToIfra
+        const childrenDataArray = await Promise.all(
+          blockIds.map(blockId => getCourseVerticalChildren(blockId)),
+        );
+        const allChildren = childrenDataArray.flatMap(data => data.children || []);
```
[optional]: Lets use `reduce` instead of `flatMap`. They have the same algorithmic complexity, but by using `reduce` we can reduce the amount of additional costs due to the lack of intermediate arrays (`map()`). If we have large amounts of data in `childrenDataArray`, the approach with `reduce` should pay for itself.

ihor-romaniuk replied:
Replaced

PKulkoRaccoonGang replied:
Resolved

PKulkoRaccoonGang commented on src/course-unit/hooks.jsx:
```diff
@@ -74,7 +79,13 @@ export const useCourseUnit = ({ courseId, blockId }) => {
     },
-    handleEdit: () => {},
+    handleEdit: () => {
+      window.postMessage({
```
[nit]: Can we send event via `sendMessageToIframe`?

ihor-romaniuk replied:
Replaced with sendMessageToIframe

PKulkoRaccoonGang replied:
Resolved

PKulkoRaccoonGang commented on src/course-unit/sidebar/SplitTestSidebarInfo.tsx:
```diff
@@ -0,0 +1,61 @@
+          {intl.formatMessage(messages.sidebarSplitTestExperimentComponentInstruction)}
+        </p>
+        <hr className="course-split-test-sidebar-devider my-4" />
+        <Link
```
[nit]: Lets use Paragon [Hyperlink](https://paragon-openedx.netlify.app/components/hyperlink/) component

PKulkoRaccoonGang replied:
Thanks, resolved

PKulkoRaccoonGang commented on src/course-unit/sidebar/SplitTestSidebarInfo.tsx:
```diff
@@ -0,0 +1,61 @@
+  );
+};
+
+SplitTestSidebarInfo.propTypes = {};
```
[question]: Do we need proptypes here?

ihor-romaniuk replied:
Removed

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-1492-ba9a65.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-1492-ba9a65.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9249653550-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9249653550-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9249653550-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-1492-ba9a65.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-1492-ba9a65.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9256377893-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9256377893-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9256377893-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-1492-ba9a65.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-1492-ba9a65.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9256968090-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9256968090-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9256968090-tutor-requirements.txt)

arbrandes reviewed (changes requested):
I haven't reviewed the code in depth, yet, but I noticed a few problems when testing in the sandbox:

* It's not possible to add an HTML ("Text") block to a content group in the new split_test page.  Clicking on it does nothing and throws no detectable error.

* Adding a problem works, but the "Select" button when selecting a problem type initially is presented too far down the page, after a blank space.  This initially led me to believe the button just wasn't showing.

* An X-Frame-Options issue?  After selecting a problem type and saving it, my browser refuses to display the contents with an "The loading of <problem block URL> in a frame is denied by “X-Frame-Options“ directive set to “sameorigin“."  Refreshing the page fixes it.  This is reproducible in this PR's sandbox.

* When copying a component with an empty clipboard, the "paste component" button does not show until after refreshing the page.  (Pasting does work afterwards, though.)

* If you click on the config link inside "This content experiment uses group configuration '[Group config 1]()' at the top of the page, the group config page is embedded in the split_test page instead of directing the user to the top-level route.

I'm not sure how many of the above issues stem from the new page.  If they're pre-existing, or if the issues are meant to be solved in a subsequent PR, we can consider merging it as is.  Let me know.

PKulkoRaccoonGang commented on src/course-unit/add-component/AddComponent.jsx:
```diff
@@ -77,126 +85,163 @@ const AddComponent = ({ blockId, handleCreateNewCourseXBlock }) => {
+          footerNode={
+            isSelectLibraryContentModalOpen && (
+              <ActionRow>
+                <Button variant="primary" onClick={onComponentSelectionSubmit}>
```
Resolved

PKulkoRaccoonGang commented on src/course-unit/context/iFrameContext.tsx:
```diff
@@ -16,11 +16,12 @@ export const IframeProvider: React.FC = ({ children }: { children: ReactNode })
   }, []);
 
-  const sendMessageToIframe = useCallback((messageType: string, payload: any) => {
+  const sendMessageToIframe = useCallback((messageType: string, payload: any, consumerWindow?: any) => {
```
[curious]: Since we have a full understanding of what's in the payload, should we type the payload structure (`type: string`, `payload: {}`)?

```
    window.parent.postMessage(
        {
            type: '<TYPE_NAME>',
            payload: { ... }
        }, document.referrer
    );
```

ihor-romaniuk replied:
We don't pass `description` to messages to iframe, so I think it's not necessary to add it here.
We pass this value from legacy to MFE but not the other way around.

PKulkoRaccoonGang replied:
Resolved

PKulkoRaccoonGang commented on src/course-unit/xblock-container-iframe/index.scss:
```diff
@@ -0,0 +1,4 @@
+.xblock-container-iframe {
+    width: calc(100% + ($spacer * .3125));
+    margin: 0 (-($spacer * .3125));
```
[optional]: Since this MFE will be moving to the use of Paragon Design tokens, it would be great to already envisage the use of CSS variables in such mathematical constructions.

```suggestion
    margin: 0 (($spacer * .3125) * -1);
```

PKulkoRaccoonGang replied:
Resolved

ihor-romaniuk commented on src/course-unit/context/iFrameContext.tsx:
```diff
@@ -5,7 +5,7 @@ import { logError } from '@edx/frontend-platform/logging';
 export interface IframeContextType {
   setIframeRef: (ref: MutableRefObject<HTMLIFrameElement | null>) => void;
-  sendMessageToIframe: (messageType: string, payload: unknown) => void;
+  sendMessageToIframe: (messageType: string, payload: unknown, consumerWindow?: any) => void;
```
Changed

PKulkoRaccoonGang replied:
Resolved

ihor-romaniuk commented:
@arbrandes Thanks for your quick review
- Fix: Adding an HTML didn't work on old legacy functional but now it should work.
- Fix: Adding a problem: implemented redirect to MFE editor (problem/video/html) page as it implemented in general unit page.
- Fix: An X-Frame-Options issue: also was fixed by implementing redirect to MFE editor (problem/video/html) page.
- Fix: copying functionality.
- Fix: fixed opening the group configuration page by routing user to the group configuration page to click on the config link inside "This content experiment uses group configuration 'Group config 1'" at the top of the page.

In additional, we have an issue with backend redirection:
- When we add new HTML, Problem, Video and automatically redirect to new MFE page we don't return to MFE Split_Test unit page. Should we fix it in this PR or postpone for future tasks?

PKulkoRaccoonGang commented on src/course-unit/add-component/AddComponent.jsx:
```diff
@@ -77,126 +83,145 @@ const AddComponent = ({ blockId, handleCreateNewCourseXBlock }) => {
-        });
+        handleCreateNewCourseXBlock(
+          { type, boilerplate: moduleName, parentLocator: blockId },
+          ({ courseKey, locator }) => navigate(`/course/${courseKey}/editor/html/${locator}`),
```
Resolved

ihor-romaniuk commented:
@arbrandes I attached the Split Test redirect issue recording and added a little more information and a possible solution:

https://github.com/user-attachments/assets/35ae67dc-cc07-4753-b8db-956785498bdf

**Issue Description:**
When users add or edit xBlocks (HTML, problem, video) using the new MFE Editor, they are incorrectly redirected to a specific group page within the current Split Test instead of returning to the Split Test page after saving or closing the editor.

**Proposed solution:**
Ensure that users are consistently redirected back to the Split Test page when exiting the MFE Editor. This adjustment will improve navigation flow, maintain user expectations, and enhance the overall experience.

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-1492-ba9a65.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-1492-ba9a65.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9372683364-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9372683364-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9372683364-tutor-requirements.txt)

arbrandes commented:
@ihor-romaniuk, thank you for the fixes!

>     * Fix: adding an HTML didn't work on old legacy functional but now it should work.

Confirmed working!

>     * Fix: adding a problem: implemented redirect to MFE editor (problem/video/html) page as it implemented in general unit page.

Works!

>     * Fix: An X-Frame-Options issue: also was fixed by implementing redirect to MFE editor (problem/video/html) page.

Confirmed!

>     * Fix: copying functionality.

Confirmed working!  The paste button shows immediately after copying if the clipboard was previously empty.
 
>     * Fix: opening the group configuration page fixed by routing user to the group configuration page to click on the config link inside "This content experiment uses group configuration 'Group config 1'" at the top of the page.

Confirmed!

> In additional, we have an issue with backend redirection:
> 
>     * When we add new HTML, Problem, Video and automatically redirect to new MFE page we don't return to MFE Split_Test or Library unit page. Should we fix it in this PR or postpone for future tasks?

The behavior I currently see in this PR sandbox is:

1. Adding or editing a block of one of the standard types in the new split_test editor correctly takes me to the correct editor page, as expected.
2. After clicking on "Save" or "Cancel", I'm taken to the legacy Studio editor of the split_test.  I expected to be taken to the new split_test editing page in the authoring MFE instead.

More on this as a reply to your subsequent post.

arbrandes commented:
> Ensure that users are consistently redirected back to the Split Test page when exiting the MFE Editor. This adjustment will improve navigation flow, maintain user expectations, and enhance the overall experience.

I agree this is what should happen, and if at all possible, I think it should be implemented in this PR.

If for any reason it would be easier to implement in a subsequent PR, that's also fine, as long as we do it as part of FC-0070.

UvgenGen commented:
@arbrandes I added new commit to https://github.com/openedx/edx-platform/pull/35813 with fix for split test and library content redirect issues (https://github.com/openedx/frontend-app-authoring/pull/1492#issuecomment-2713341098).
Please take a look. I will be glad for comments!

arbrandes reviewed (approved):
Ok, this is working great, as far as I tested.  Thanks for bearing with the reviews and addressing all issues!

arbrandes commented:
This is good to merge once the tests pass.  At the moment it seems there's only a few linting failures.

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9585113967.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9585113967-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9585113967-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9585113967-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9589901180.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9589901180-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9589901180-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-1492-ba9a65-51564656-9589901180-tutor-requirements.txt)

arbrandes commented:
I ran the tests locally as well, and they check out.  We have a slight reduction in coverage, but it is small enough that we can let it slide.

arbrandes commented:
Related PR to tutor-mfe, for reference:

https://github.com/overhangio/tutor-mfe/pull/249
